File: baekjoon/15000/15481.cpp
#include <bits/stdc++.h>
using namespace std;

struct edge {
    int u, v, w, id;
};

struct disjoint_set {
    vector<int> par, dist;
    disjoint_set(int n) : par(n, -1), dist(n, 0x3f3f3f3f) {}
    int find(int u) {
        while (par[u] >= 0) u = par[u];
        return u;
    }
    bool merge(int u, int v, int w) {
        u = find(u), v = find(v);
        if (u == v) return false;

        if (par[u] > par[v]) swap(u, v);
        par[u] += par[v];
        par[v] = u;

        dist[v] = w;
        return true;
    }
    int query(int u, int v) {
        int ret;
        for (; u != v; u = par[u]) {
            if (dist[u] > dist[v]) swap(u, v);
            ret = dist[u];
        }
        return ret;
    }
};

int main() {
    cin.tie(nullptr)->sync_with_stdio(false);
#ifdef home
    freopen("in", "r", stdin);
    freopen("out", "w", stdout);
#endif
    int n, m;
    cin >> n >> m;

    vector<edge> e(m);
    for (int i = 0; i < m; ++i) {
        cin >> e[i].u >> e[i].v >> e[i].w;
        --e[i].u, --e[i].v, e[i].id = i;
    }

    sort(e.begin(), e.end(), [&](auto& a, auto& b) {
        return a.w < b.w;
    });

    vector<int> ans(m, -1);
    disjoint_set dsu(n);
    int64_t mst = 0;

    for (const auto& [u, v, w, id] : e)
        if (dsu.merge(u, v, w))
            mst += w, ans[id] = 0;

    for (const auto& [u, v, w, id] : e)
        if (ans[id] == -1)
            ans[id] = w - dsu.query(u, v);

    for (const auto& x : ans)
        cout << mst + x << '\n';
}

File: baekjoon/15000/15481.rs
use std::io::*;

struct Scanner<'a> {
    it: std::str::SplitWhitespace<'a>,
}

impl<'a> Scanner<'a> {
    fn new(s: &'a str) -> Scanner<'a> {
        Scanner {
            it: s.split_whitespace(),
        }
    }
    fn read<T: std::str::FromStr>(&mut self) -> T {
        self.it.next().unwrap().parse::<T>().ok().unwrap()
    }
}

fn main() {
    let mut s = String::new();
    stdin().read_to_string(&mut s).unwrap();
    let mut sc = Scanner::new(&s);
    let out = stdout();
    let mut out = BufWriter::new(out.lock());
    run(&mut sc, &mut out);
}

struct DisjointSet {
    par: Box<[i32]>,
    dist: Box<[u64]>,
}

impl DisjointSet {
    fn new(n: usize) -> Self {
        Self {
            par: vec![-1; n].into_boxed_slice(),
            dist: vec![u64::MAX; n].into_boxed_slice(),
        }
    }
    fn find(&self, mut u: usize) -> usize {
        while self.par[u] >= 0 {
            u = self.par[u] as usize;
        }
        u
    }
    fn unite(&mut self, mut u: usize, mut v: usize, w: u64) -> bool {
        u = self.find(u);
        v = self.find(v);
        if u == v {
            return false;
        }
        if self.par[u] > self.par[v] {
            std::mem::swap(&mut u, &mut v);
        }
        self.par[u] += self.par[v];
        self.par[v] = u as i32;
        self.dist[v] = w;
        true
    }
    fn query(&self, mut u: usize, mut v: usize) -> u64 {
        let mut ret = u64::MAX;
        while u != v {
            if self.dist[u] > self.dist[v] {
                std::mem::swap(&mut u, &mut v);
            }
            ret = self.dist[u];
            u = self.par[u] as usize;
        }
        ret
    }
}

fn run<W: Write>(sc: &mut Scanner, out: &mut BufWriter<W>) {
    let n = sc.read::<usize>();
    let m = sc.read::<usize>();
    let mut edges = (0..m)
        .map(|i| {
            (
                sc.read::<usize>() - 1,
                sc.read::<usize>() - 1,
                sc.read::<u64>(),
                i,
            )
        })
        .collect::<Vec<_>>();
    edges.sort_unstable_by(|l, r| l.2.cmp(&r.2));
    let mut dsu = DisjointSet::new(n);
    let mut mst = 0;
    for &(u, v, w, _) in edges.iter() {
        if dsu.unite(u, v, w) {
            mst += w;
        }
    }
    let mut ans = vec![0; m];
    for (u, v, w, i) in edges.into_iter() {
        ans[i] = mst + w - dsu.query(u, v);
    }
    for x in ans {
        writeln!(out, "{}", x).ok();
    }
}
